protected/humhub/services/MigrationService.php

```diff
@@ -0,0 +1,262 @@
+<?php
+
+/*
+ * @link      https://www.humhub.org/
+ * @copyright Copyright (c) 2023 HumHub GmbH & Co. KG
+ * @license   https://www.humhub.com/licences
+ */
+
+namespace humhub\services;
+
+use humhub\commands\MigrateController;
+use humhub\components\Event;
+use humhub\components\Module;
+use humhub\events\MigrationEvent;
+use Throwable;
+use Yii;
+use yii\base\ActionEvent;
+use yii\base\Component;
+use yii\base\Controller;
+use yii\base\InvalidConfigException;
+use yii\console\ExitCode;
+
+/**
+ * @since 1.16
+ */
+class MigrationService extends Component
+{
+    public const EVENT_AFTER_MIGRATION = 'afterMigration';
+    public const EVENT_BEFORE_MIGRATION = 'beforeMigration';
+    public const ACTION_ENABLE = 'enable';
+    public const ACTION_DISABLE = 'disable';
+    public const ACTION_UPDATE = 'update';
+    public const ACTION_MIGRATE = 'migrate';
+    protected const MIGRATION_UNINSTALL = 'uninstall';
+    protected const MIGRATION_UP = 'up';
+
+    protected Module $module;
+    protected ?string $path;
+
+    /**
+     * @param Module $module
+     * @param array $config = ['path' => string|null]
+     */
+    public function __construct(Module $module, array $config = [])
+    {
+        $this->module = $module;
+
+        $path = $config['path'] ?? null;
```

Comment:
I'm not 100% sure about the paths in relation to modules:

- the module's `config.php` has to reside in a subfolder of any of the module search path (app config parameter).
- however, the module can be _anywhere_, as long as one registered auto-loader can find it.

My questions:
1. Is now the `/migrations` folder always relative to the `config.php` file or the module's class file (usually, but not necessarily `module.php`)?
2. Are there real-world situations where the two paths (of `config.php` and the module class file) are different?

And yes, if we use the reflection to get the module's path, the class will be loaded. And yes, this might have been an issue.

But I think we should consider to revisit the `MigrationService` together with the `ModuleService` that I have already written to do the install, update, remove tasks of the module. I find it a bit difficult to discus details that do depend on code that is not yet in this PR (to keep it in an manageable size). If you prefer, we could apply this PR to a new branch and then apply the other two on top of it: a) for renaming `active` => `enabled`, and b) the `ModuleService`.

> Maybe we could make $module optional, when not given, the migration service covers alls migrations?

I like that idea. I think we could also move some functionality from the `MigrateController` to the `MigrateService`, keeping the "routing and managing" in the controller, but the `executing` in the service. And then either pass the application as the `$module` parameter or leave it optional which will default to `Yii::$app` ...

Reply:
> I'm not 100% sure about the paths in relation to modules:
> 
> * the module's `config.php` has to reside in a subfolder of any of the module search path (app config parameter).
> * however, the module can be _anywhere_, as long as one registered auto-loader can find it.
> 
> My questions:
> 
> 1. Is now the `/migrations` folder always relative to the `config.php` file or the module's class file (usually, but not necessarily `module.php`)?

Following structure should be expected: https://docs.humhub.org/docs/develop/modules#module-structure

A HumHub module can have a `migrations` folder in its base folder.
The same folder also contains the module class `Module.php` and `config.php`.    

I would not assume special cases here.

> 2. Are there real-world situations where the two paths (of `config.php` and the module class file) are different?

Not sure, what you mean. 

But `config.php` and `Module.php` file names are fixed. Theoretically, `Module.php` could have a different class name. However, I am not aware of any cases.
> 
> And yes, if we use the reflection to get the module's path, the class will be loaded. And yes, this might have been an issue.



> 
> But I think we should consider to revisit the `MigrationService` together with the `ModuleService` that I have already written to do the install, update, remove tasks of the module. I find it a bit difficult to discus details that do depend on code that is not yet in this PR (to keep it in an manageable size). If you prefer, we could apply this PR to a new branch and then apply the other two on top of it: a) for renaming `active` => `enabled`, and b) the `ModuleService`.
> 

Unfortunately, a tricky situation. I struggle a lot with reviewing large PR changes and I'm concerned that potentially small changes will be not discussed in detail. 

> > Maybe we could make $module optional, when not given, the migration service covers alls migrations?
> 
> I like that idea. I think we could also move some functionality from the `MigrateController` to the `MigrateService`, keeping the "routing and managing" in the controller, but the `executing` in the service. And then either pass the application as the `$module` parameter or leave it optional which will default to `Yii::$app` ...

Reply:
> Is it really worth implementing a `$config['path']` variable in addition to the module?

I have removed the explicit evaluation of the path variable. But given it's a component, a calling method could still set the path upon initialization. Which, in fact. is used for backward-compatibility of the `MigrateController::webMigration()`. However, by default the path is now evaluated by the module's path alias "@$module-id/migrations" (or "@humhub/migrations" for the application itself).